### PR TITLE
Factor publishing into its own file

### PR DESCRIPTION
**Pull request by janfreyberg**
This allows for better testing

### CODE REVIEW

File: faculty_mill/publish.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
+import os
+from pathlib import Path
+from typing import Optional
+from uuid import UUID
+from time import sleep
+import click
```
Comment (liamcoatman) [inline faculty_mill/publish.py]: Add blank line between built-in and 3rd party libraries

File: faculty_mill/publish.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    show_code: bool = False,
+    report_id: Optional[UUID] = None,
+    project_id: Optional[UUID] = None,
+    user_id: Optional[UUID] = None,
+):
+    """Publish a notebook as a report.
+
+    Parameters
+    ----------
+    report_name : str
```
Comment (liamcoatman) [inline faculty_mill/publish.py]: Is there a reason to include the types in the docstring if we are typing the variables?

File: faculty_mill/publish.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    project_id : Optional[UUID], optional
+        The project ID. Only needed if not invoking from within a project.
+    user_id : Optional[UUID], optional
+        The user ID. Only needed if not invoking from within a project.
+    show_code : bool, optional
+        Whether the code should be shown in the report or not (default False)
+    """
+
+    if project_id is None:
+        project_id = UUID(os.getenv("SHERLOCKML_PROJECT_ID"))
```
Comment (liamcoatman) [inline faculty_mill/publish.py]: I prefer `project_id  = project_id or UUID(os.getenv("SHERLOCKML_PROJECT_ID"))`

File: faculty_mill/publish.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+    if user_id is None:
+        user_id = sml.auth.user_id()
```
Comment (liamcoatman) [inline faculty_mill/publish.py]: Same here: `user_id = user_id or sml.auth.user_id()`

File: faculty_mill/publish.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
+import os
+from pathlib import Path
+from typing import Optional
+from uuid import UUID
+from time import sleep
```
Comment (liamcoatman) [inline faculty_mill/publish.py]: I prefer importing `time`, and then using `time.sleep()` for its explicitness

File: faculty_mill/publish.py (hunk 2 above)
Comment (janfreyberg) [reply faculty_mill/publish.py]: I think for discoverability?

Comment (liamcoatman) [reply faculty_mill/publish.py]: What do you mean?

Comment (janfreyberg) [reply faculty_mill/publish.py]: Some people might only look at the doc string to find the types, so this would help them

Comment (liamcoatman) [reply faculty_mill/publish.py]: Fair enough

File: faculty_mill/publish.py (hunk 4 above)
Comment (srstevenson) [reply faculty_mill/publish.py]: Ditto.

File: faculty_mill/publish.py (hunk 3 above)
Comment (janfreyberg) [reply faculty_mill/publish.py]: this is why I had it initially, but in this case it's either a UUID or None, so it seems safe

File: faculty_mill/publish.py (hunk 1 above)
Comment (srstevenson) [reply faculty_mill/publish.py]: If we're going to adopt a style of keeping imports sorted (which we should!), let's setup `isort` so we can automate it and not need to think about it again.

File: faculty_mill/publish.py (hunk 3 above)
Comment (srstevenson) [reply faculty_mill/publish.py]: I think in the spirit of "Explicit is better than implicit", and being consistent so we don't get in the habit of using the shortcut in contexts where it does behave differently, the `is None` check is better.

Comment (acroz) [reply faculty_mill/publish.py]: Given there's the `Optional[UUID]` type hint, I think either is acceptable, however I prefer the clarity that comes with the ` if _ is None:` statement.

I know that mypy is able to use the `if project_id is None` test to know that `project_id` is a `UUID` and not an `Optional[UUID]` on subsequent lines - I'm not sure if that is also True under @liamcoatman's proposed change.

File: tests/test_publish.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+        result = publish.get_report_id_by_name("test name", "project id")
+        mock_client_creator.assert_called_once_with("report")
+        mock_client.list.assert_called_once_with("project id")
+        assert result == mock_report.id
+
+
+def test_that_publish_calls_client_method_correctly_with_all_ids_set():
+
+    mock_client = Mock()
```
Comment (acroz) [inline tests/test_publish.py]: I suggest using `pytest-mock` as we do in the sherlockml library. This provides a pytest fixture that unapplies the mock automatically after the test, avoiding the need to indent most of the content of tests in context managers.

Have a look at the sherlockml library tests for examples of how to use it.

File: faculty_mill/publish.py (hunk 3 above)
Comment (janfreyberg) [reply faculty_mill/publish.py]: In fact, mypy can't use `if something` to disambiguate `None`'s, so that's probably a good reason to be explicit

Comment (zblz) [reply faculty_mill/publish.py]: With regards to @srstevenson original concern testing whether the argument is falsey, we definitely don't want to allow a `project_id` argument value like `False` to be overridden from the environment. IMO `if _ is None` is definitely needed.

File: faculty_mill/publish.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+import click
+import sherlockml
+import sml.auth
```
Comment (zblz) [inline faculty_mill/publish.py]: `sml.auth.user_id()` should be replaced with the `AccountClient` in `sherlockml`. In general, we should avoid importing `sml` in other libraries.

File: tests/test_publish.py (hunk 6 above)
Comment (janfreyberg) [reply tests/test_publish.py]: I decided to go for the decorator option instead, as that seems a bit more obvious and avoids extra dependencies / fixture magics

File: faculty_mill/publish.py (hunk 1 above)
Comment (janfreyberg) [reply faculty_mill/publish.py]: added a few pre-commit hooks for this exact purpose

**Comment by janfreyberg** [comment]
All fixed!

**Comment by janfreyberg** [comment]
Hmmmmm I don't know why coveralls is not reporting anything

File: .pre-commit-config.yaml [hunk 8]
```diff
@@ -0,0 +1,13 @@
+repos:
+-   repo: https://github.com/pre-commit/mirrors-isort
+    rev: HEAD  # Use the revision sha / tag you want to point at
+    hooks:
+    -   id: isort
```
Comment (srstevenson) [inline .pre-commit-config.yaml]: There's two extra spaces between `-` and `id` here.

File: tests/test_publish.py (hunk 6 above)
Comment (acroz) [reply tests/test_publish.py]: Well it's nice as you can then write your mocks as pytest fixtures and share them. Pytest fixtures are unpleasantly magic for Python, but they save a lot of boilerplate compared to standard unittest constructs.

There is arguably also value in some level of consistency in our published Python code - we use pytest-mock in most Python code in the SherlockML team, both internally and in the sherlockml library.

**Comment by janfreyberg** [comment]
> Hmmmmm I don't know why coveralls is not reporting anything

I think there's a problem with coveralls at the moment.

File: tests/test_publish.py (hunk 6 above)
Comment (janfreyberg) [reply tests/test_publish.py]: Yeah - I use it on superintendent, as well, but it feels icky :P If we get more tests that need this in the future, we can always implement it in a separate PR.

**Pull request merged**